src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+                    .invert_selection_fg_bg = config.@"selection-invert-fg-bg",
+                    .bold_is_bright = config.@"bold-is-bright",
+                    .min_contrast = @floatCast(config.@"minimum-contrast"),
+                    .padding_color = config.@"window-padding-color",
+
+                    .selection_background = if (config.@"selection-background") |bg|
+                        bg.toTerminalRGB()
+                    else
+                        null,
+
+                    .selection_foreground = if (config.@"selection-foreground") |bg|
+                        bg.toTerminalRGB()
+                    else
+                        null,
+
+                    .custom_shaders = custom_shaders,
+                    .links = links,
+                    .vsync = config.@"window-vsync",
+                    .colorspace = config.@"window-colorspace",
+                    .blending = config.@"alpha-blending",
+                    .arena = arena,
+                };
+            }
+
+            pub fn deinit(self: *DerivedConfig) void {
+                const alloc = self.arena.allocator();
+                self.links.deinit(alloc);
+                self.arena.deinit();
+            }
+        };
+
+        /// Returns the hints that we want for this window.
+        pub fn glfwWindowHints(config: *const configpkg.Config) glfw.Window.Hints {
+            // If our graphics API provides hints, use them,
+            // otherwise fall back to generic hints.
+            if (@hasDecl(GraphicsAPI, "glfwWindowHints")) {
+                return GraphicsAPI.glfwWindowHints(config);
+            }
+
+            return .{
+                .client_api = .no_api,
+                .transparent_framebuffer = config.@"background-opacity" < 1,
+            };
+        }
+
+        pub fn init(alloc: Allocator, options: renderer.Options) !Self {
+            // Initialize our graphics API wrapper, this will prepare the
+            // surface provided by the apprt and set up any API-specific
+            // GPU resources.
+            var api = try GraphicsAPI.init(alloc, options);
+            errdefer api.deinit();
+
+            const has_custom_shaders = options.config.custom_shaders.value.items.len > 0;
+
+            // Prepare our swap chain
+            var swap_chain = try SwapChain.init(
+                api,
+                has_custom_shaders,
+            );
+            errdefer swap_chain.deinit();
+
+            // Create the font shaper.
+            var font_shaper = try font.Shaper.init(alloc, .{
+                .features = options.config.font_features.items,
+            });
+            errdefer font_shaper.deinit();
+
+            // Initialize all the data that requires a critical font section.
+            const font_critical: struct {
+                metrics: font.Metrics,
+            } = font_critical: {
+                const grid = options.font_grid;
```

Comment:
I know this is copied from old code so its not your fault, but we should put an explicit type here because we want to be sure we're always getting a pointer and never a copy (if we ever change the options types).